File: leetcode/hard/escape_large_maze.py
class Solution:
    def isEscapePossible(self, blocked: List[List[int]], source: List[int], target: List[int]) -> bool:
        dirs = ((0,1), (1,0), (-1,-1), (1,-1),
                (0,-1), (-1,0), (1,1), (-1,1))

        def addCX(x, y):
            chainX[x][0] = min(chainX[x][0], y)
            chainX[x][1] = max(chainX[x][1], y)
        def addCY(x, y):
            chainY[y][0] = min(chainY[y][0], x)
            chainY[y][1] = max(chainY[y][1], x)

        def contained(x, y):
            ya, yb = chainX
            xa, xb = chainY
            return ya <= y <= yb and xa <= x <= xb

        def dfs(x, y):
            addCX(x, y), addCY(x, y)
            seen.add( (x,y) )
            for d in dirs:
                r, c = x+d[0], y+d[1]
                if (r,c) in obs and (r,c) not in seen:
                    dfs(r, c)

        # if # of blocks NOT limited to 200:
        # maintain running region
        # if (x,y) in seen
        #   region = list[idx[seen] : -1]
        #   check if src/target contained,
        # count number of regions containing src/target
        # if they match then true
        # 2 axis hits: axes become bounds

        seen = set()
        sx, sy = source
        tx, ty = target
        obs = set(tuple(b) for b in blocked)
        for x,y in blocked:
            chainX = defaultdict(lambda: [float('inf'), -float('inf')])
            chainY = defaultdict(lambda: [float('inf'), -float('inf')])
            dfs(x,y)
            if contained(sx, sy) and not contained(tx, ty):
                return False

File: leetcode/hard/escapeLargeMaze.cpp
// # if # of blocks NOT limited to 200:
// # maintain running region
// # if (x,y) in seen
// #   region = list[idx[seen] : -1]
// #   check if src/target contained,
// # count number of regions containing src/target
// # if they match then true
// # 2 axis hits: axes become bounds
// sad that this convoluted mess wasn't the real solution lol
class Solution {
public:
    unordered_map<int, unordered_set<int>> obs;
    int dirs[4][2] = {{0,1}, {1,0}, {-1,0}, {0,-1}};

    bool bfs(int sx, int sy, int tx, int ty) {
        queue<pair<int, int>> q;
        q.push({sx, sy});
        unordered_map<int, unordered_set<int>> seen;
        int size = 0;
        while (!q.empty()) {
            auto [x,y] = q.front();
            q.pop();
            if (x == tx && y == ty) return true;
            if (size >= 19900) return true;
            if (seen[x].count(y)) continue;
            seen[x].insert(y);
            ++size;
            for (auto& d : dirs) {
                int r = x + d[0];
                int c = y + d[1];
                if (seen[r].count(c)) continue;
                if (r < 0 || r >= 1000000) continue;
                if (c < 0 || c >= 1000000) continue;
                if (obs[r].count(c)) continue;
                q.push({r, c});
            }
        }
        return false;
    }

    bool isEscapePossible(vector<vector<int>>& blocked, vector<int>& source, vector<int>& target) {
        int sx = source[0], sy = source[1];
        int tx = target[0], ty = target[1];
        for (auto& b : blocked) obs[b[0]].insert(b[1]);
        return bfs(sx, sy, tx, ty) && bfs(tx, ty, sx, sy);
    }
};
